### PR TITLE
disable color-eyre colors if not connected to a tty

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

Prior to this change we applied some fixes to let us suppress ansi code output from zebrad so we could still read the logs after they'd been piped to a file. These changes had to stop short of color-eyre because it had no option to disable colors, and disabling it entirely when colors are annoying was worse than just ignoring the ansi codes in a file. 

## Solution

This PR utilizes the new custom theme support of color-eyre to set a blank theme when we detect that either stdout or stderr isn't a tty. 

The code in this pull request has:
  - [ ] Documentation Comments
  - [ ] Unit Tests and Property Tests

### CODE REVIEW

**Review by dconnolly — DISMISSED** [review]
yaaaay this automatically works in the gcloud log parsing 🎉

image

**Review by teor2345 — CHANGES_REQUESTED** [review]
We missed this part of the change:
* add a configuration that always turns colors off, because auto-detection isn't perfect, some users won't have color terminals, and others will want colors off for accessibility

We already have `config.tracing.use_color`. If we install the panic handler after the config is loaded, then we can make it respect the user's color config.

**Comment by teor2345** [comment]
We could also disable colors in `color-backtrace` by:
1. Creating a `color_backtrace::ColorScheme` where each field is `termcolor::ColorSpec::default()`
2. Calling `BacktracePrinter::default().color_scheme(no_colors).install(default_output_stream())`

We need to make our own `ColorScheme`, because `color_backtrace::ColorScheme::default()` is the `classic()` colored scheme.

https://docs.rs/color-backtrace/0.5.0/color_backtrace/struct.ColorScheme.html
https://docs.rs/color-backtrace/0.5.0/color_backtrace/struct.BacktracePrinter.html#method.color_scheme

Both these changes should be quick, and it's good to get them right while we're working on this PR, then we can focus on other stuff.

**Comment by yaahc** [comment]
> We could also disable colors in `color-backtrace` by:
> 
>     1. Creating a `color_backtrace::ColorScheme` where each field is `termcolor::ColorSpec::default()`
> 
>     2. Calling `BacktracePrinter::default().color_scheme(no_colors).install(default_output_stream())`
> 
> 
> We need to make our own `ColorScheme`, because `color_backtrace::ColorScheme::default()` is the `classic()` colored scheme.
> 
> https://docs.rs/color-backtrace/0.5.0/color_backtrace/struct.ColorScheme.html
> https://docs.rs/color-backtrace/0.5.0/color_backtrace/struct.BacktracePrinter.html#method.color_scheme
> 
> Both these changes should be quick, and it's good to get them right while we're working on this PR, then we can focus on other stuff.

Color backtrace should never actually get used in our application. It installs a hook which we then immediately replace via color-eyre, throwing away the color-backtrace hook in the process.

**Comment by yaahc** [comment]
> We already have `config.tracing.use_color`. If we install the panic handler after the config is loaded, then we can make it respect the user's color config.

I held off on doing this because I recalled that you _had_ done so just a day or two ago, and it looked like it had been reverted or updated from when I last reviewed it and was moved back to before configuration was loaded. Am I just misremembering or did it get changed? I was worried that there was some regression when the ordering was changed that I'd have to worry about.

**Review by dconnolly — DISMISSED** [review]

**Comment by teor2345** [comment]
> > We already have `config.tracing.use_color`. If we install the panic handler after the config is loaded, then we can make it respect the user's color config.
> 
> I held off on doing this because I recalled that you _had_ done so just a day or two ago, and it looked like it had been reverted or updated from when I last reviewed it and was moved back to before configuration was loaded. Am I just misremembering or did it get changed? I was worried that there was some regression when the ordering was changed that I'd have to worry about.

I backed out the changes because you said that disabling `color-eyre` was worse than having it write colors to a file, and I wanted to get PR #1412 merged. There are no regressions.
https://github.com/ZcashFoundation/zebra/pull/1412#issuecomment-737008192

I'll try to be clearer next time - and if you're skipping changes, please mention them in the follow-up section of the PR.

File: zebrad/src/application.rs [hunk 1]
```diff
@@ -45,6 +45,10 @@ pub struct ZebradApp {
 }
 
 impl ZebradApp {
+    fn is_tty() -> bool {
```
Comment (teor2345) [inline zebrad/src/application.rs]: Unix has 3 standard streams, input, output, and error. TTYs can also be attached to other file descriptors.
```suggestion
    /// Are standard output and standard error both connected to ttys?
    fn outputs_are_ttys() -> bool {
```

File: zebrad/src/application.rs [hunk 2]
```diff
@@ -109,30 +113,12 @@ impl Application for ZebradApp {
             //
             // We'd also like to check `config.tracing.use_color` here, but the
             // config has not been loaded yet.
-            if !atty::is(atty::Stream::Stdout) || !atty::is(atty::Stream::Stderr) {
+            if !Self::is_tty() {
```
Comment (teor2345) [inline zebrad/src/application.rs]:
```suggestion
            if !Self::outputs_are_ttys() {
```

**Review by teor2345 — APPROVED** [review]
This looks good now, let's merge after it passes CI

**Pull request merged**